3scale/3scale-operator
# THREESCALE-10238 - mapping rules reconciliation improvement

## CR will be closed. Please see PR https://github.com/3scale/3scale-operator/pull/924  - suggested for review and approval.
Jira: https://issues.redhat.com/browse/THREESCALE-10238

- Please see detailed description and related tasks in [THREESCALE-10238](https://issues.redhat.com/browse/THREESCALE-10238)

# Example of the problem that fixed in current PR

- **Product CR contains "duplicated" rules:**
![productCRDUplicatedRules](https://github.com/3scale/3scale-operator/assets/56625811/8388c11a-b359-4377-a0aa-ec4e9225d444)

## Mapping Rules in portal Before Fix 
![beforeFix](https://github.com/3scale/3scale-operator/assets/56625811/694ee292-7952-4217-9ffe-71f11a286bcb)

## Mapping Rules in portal After Fix in current PR
![afterFix](https://github.com/3scale/3scale-operator/assets/56625811/220b5cd7-7a11-4935-b46f-d6cc2f08e407)

## Add new Rule in CR:
![Screenshot from 2024-01-25 10-09-11](https://github.com/3scale/3scale-operator/assets/56625811/2f21d28b-c80a-41f5-b0ba-57019a391d8e)

![Screenshot from 2024-01-25 10-08-33](https://github.com/3scale/3scale-operator/assets/56625811/960d9bfd-1c9f-48b2-8884-e81a83c2e8f6)


##  Problem to be resolved  -TODO ?
- **can't add "duplicated" rule in CR** . 
 Can't add rule if identical rule exists. Although rule can be edited (increment field can be changed)
Need investigate the problem and allow this.

- when upgrade to new 3scale Release (with PR)  - Not valid Old rules remains and need to be cleaned manually (?)
![problemAferFixExistedRules](https://github.com/3scale/3scale-operator/assets/56625811/12436825-e876-4b4f-9639-861105cd61c8)

 ## Validation 
 <details>

You can try compare the behavior on muster branch (before fix), and after fix (current PR)

```
make install
oc new-project 3scale-test
make download
make run
```

- On another window:
Apply following files. The examples for backend and product - provided below. Other - please ask me if required.
```
oc apply -f s3-creds-secret.yaml
oc apply -f apimanagerCR.yaml
oc apply -f backend.yml
oc apply -f product.yml
```
- Check mapping rules, compare with Product and Backend CRs
- Try add rules (uniq and identical)


### Product and Backend files examples
file contains duplicated rules to show the problem

```yaml
apiVersion: capabilities.3scale.net/v1beta1
kind: Product
metadata:
  name: product1
spec:
  name: "OperatedProduct 1"
  systemName: "operatedproduct1"
  description: "Operated Product 1 for reconciliation testing"
  backendUsages:
    backend1:
      path: /
  mappingRules:
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: method01
      pattern: /pets 
      
  metrics:
    hits:
      description: Number of API hits
      friendlyName: Hits
      unit: hit
  methods:
    method01:
      friendlyName: Method01
```

```yaml
apiVersion: capabilities.3scale.net/v1beta1
kind: Backend
metadata:
  name: backend1
spec:
  name: "Operated Backend 1"
  systemName: "backend1"
  description: "Operated Backend 1 for reconciliation testing"
  privateBaseURL: "https://example.3scale.net"
  mappingRules:
    - httpMethod: PUT
      increment: 1
      metricMethodRef: hits
      pattern: /pets
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /pets      
    - httpMethod: GET
      increment: 1
      metricMethodRef: hits
      pattern: /cars
    - httpMethod: GET
      increment: 1
      metricMethodRef: method01
      pattern: /pets
    - httpMethod: GET
      increment: 1
      metricMethodRef: method01
      pattern: /pets
  metrics:
    hits:
      description: Number of API hits
      friendlyName: Hits
      unit: hit
  methods:
    method01:
      friendlyName: Method01

```

 </details>
